### ISSUE TITLE
PyQt5 to heavy weight compilation

### ISSUE THREAD

**Issue by AgustinPardo** [pyinstaller/pyinstaller#5503]
Hello,

I am compiling a python little app. This app imports PyQt5 modules.

The modules are:

```python
from PyQt5.QtWidgets import QApplication, QWidget, QPushButton, QLabel, QGridLayout, QTextEdit, QLineEdit
from PyQt5.QtCore import QSize
```

When I include the modules in the compilation the final size is 64 Mb (The execution of the binary is super slow and have lag)

When I take off the complete PyQt5 module, the final size is about 4 Mb (But of course is not working).

Do you know hoy I could use the "--exclude-module" flag to get rid of the PyQt5 modules that I am not using?

Regards

**Comment by rokm** [comment]
> Do you know hoy I could use the "--exclude-module" flag to get rid of the PyQt5 modules that I am not using?

PyInstaller is trying to include only the modules you are using (and their dependencies, of course). 

However, it currently does collect all plugins for a certain module, which results in additional dependencies being pulled in (e.g., `platforms/qwebgl` plugin ends up pulling in `Qt5DBus`, `Qt5Quick`, `Qt5Qml`, `Qt5WebSocket`, `Qt5Network`, etc.). `--exclude-module` won't work here, because these shared libraries are pulled in by linked-library analysis. 

I don't think there's currently anything you can do about that except for manually removing these files from the dist folder (works only for `onedir`).

Or, you could remove all unused plugins from the PyQt5 installation that you use for freezing the program.

> The execution of the binary is super slow and have lag

I take it you are building in `onefile` mode and are talking about the initial unpacking when the application is run?

**Comment by BoboTiG** [comment]
Yep, manual cleanup is the only solution for now. Here is what I am using, it may be useful: [cleanup_application_tree.py](https://github.com/nuxeo/nuxeo-drive/blob/master/tools/cleanup_application_tree.py).

**Comment by AgustinPardo** [comment]
@rokm 

> I don't think there's currently anything you can do about that except for manually removing these files from the dist folder (works only for onedir).

Are there common PyQt5 libraries that always import and are not used?

> Or, you could remove all unused plugins from the PyQt5 installation that you use for freezing the program.

How to do that? You mean remove them in the /dist folder?

> I take it you are building in onefile mode and are talking about the initial unpacking when the application is run?

I mean the execution of the compiled file ("onefile" or  non one file compilation is the same) is very slow, very annoying, and not useful

@BoboTiG 
You mean to run that script after no "onefile" compilation and remove them in the dist folder?

Many regards!

**Comment by 003userye** [comment]
I suggest that when you are compiling and packaging, use pyinstaller to choose manual packaging, and specify each module you call to eliminate the waste caused by packaging unnecessary modules.

Provide some specific parameters to repackage with pyinstaller, you can check its official website for specific methods

-F _After packaging, only a single exe file is generated._
-D _The default option to create a directory that contains exe files and a large number of dependent files._
-C _The default option, using the console (that is, a black box similar to cmd);_
-W _Do not use the console;_
-P _Add a search path to find the corresponding Library;_
-I _Change the icon of the generator_

**Comment by AgustinPardo** [comment]
@003userye 
Thanks. Where is the option to "manually" select the modules/libraries?

**Comment by BoboTiG** [comment]
@AgustinPardo I am running the script after PyInstaller yes. In `--onedir` mode.

**Comment by 003userye** [comment]
> @003userye 
> Thanks. Where is the option to "manuallyvolume_up
150 / 500

What I mean by ``manually'' here is that when editing the pyinstaller packaging instructions, add some parameters, and only package the modules you call according to the specific situation. For specific methods, please refer to ``https://www.cnblogs.com/bbiu/p/13209612.html'' My family is in China, and it is now 1:25.am, I have to sleep baby, good luck " select the modules/libraries?

**Comment by AgustinPardo** [comment]
@BoboTiG 
Thanks for the answer. I run the "cleaning" script. Before running the folder dist was 181 Mb, and after 175 Mb. I think it doesn't make a big difference.

It removes:

```
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqico.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqgif.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqjpeg.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqtiff.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqwbmp.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqicns.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqwebp.so
[X] Removed dist/run/PyQt5/Qt/plugins/imageformats/libqtga.so
[X] Removed dist/run/PyQt5/Qt/translations
[X] Removed dist/run/libQt5WebSockets.so.5
```
I have still have a lot of PyQt5 modules. 

Do you have another idea? How should I know which modules are critical or not for the app?

Regards

**Comment by AgustinPardo** [comment]
@003userye 
Thanks for the link. But I still don't found where is the flag to select specific modules!
Regards

**Comment by rqx110** [comment]
assume you are using pyinstaller from code, not from `pyinstaller.exe`, you can patch the `onefiletmplt` or `onedirtmplt` strings, in these template string you can remove any `datas` or `binaries`.

onefile as example:
```py
import PyInsatller.building.templates as pyi_spec_templates
from PyInstaller.__main__ import run as pyi_build

# copy from PyInsatller.building.templates.py
# add datas and binaries filter after Analysis
onefiletmplate = """# -*- mode: python ; coding: utf-8 -*-
%(preamble)s

a = Analysis(
    %(scripts)s,
    pathex=%(pathex)s,
    binaries=%(binaries)s,
    datas=%(datas)s,
    hiddenimports=%(hiddenimports)s,
    hookspath=%(hookspath)r,
    hooksconfig={},
    runtime_hooks=%(runtime_hooks)r,
    excludes=%(excludes)s,
    noarchive=%(noarchive)s,
    optimize=%(optimize)r,
)

# begin filter any datas you want
datas = []
for dest_name, src_name, res_type in a.datas:
    # add conditional judgment here
    datas.append((dest_name, src_name, res_type))
a.datas = datas
# end filter datas

# begin filter any binaries you want
binaries = []
for dest_name, src_name, res_type in a.binaries:
    # add conditional judgment here
    binaries.append((dest_name, src_name, res_type))
a.binaries = binaries
# end filter datas

pyz = PYZ(a.pure)
%(splash_init)s
exe = EXE(
    pyz,
    a.scripts,
    a.binaries,
    a.datas,%(splash_target)s%(splash_binaries)s
    %(options)s,
    name='%(name)s',
    debug=%(debug_bootloader)s,
    bootloader_ignore_signals=%(bootloader_ignore_signals)s,
    strip=%(strip)s,
    upx=%(upx)s,
    upx_exclude=%(upx_exclude)s,
    runtime_tmpdir=%(runtime_tmpdir)r,
    console=%(console)s,
    disable_windowed_traceback=%(disable_windowed_traceback)s,
    argv_emulation=%(argv_emulation)r,
    target_arch=%(target_arch)r,
    codesign_identity=%(codesign_identity)r,
    entitlements_file=%(entitlements_file)r,%(exe_options)s
)
"""

pyi_spec_templates.onefiletmplt = onefiletmplate  # ==> patch the template string here

pyi_build(....)
```

**Comment by DarkForceFREEFIRE** [comment]
I think deleting the unnecessary files and dlls from venv is essential. 
First, I executed my Qt application, then deleted many dlls and folders those aren't using for the executed Qt app.
